design/automatic-network-configuration.md

```diff
@@ -0,0 +1,485 @@
+    - name:
+      namespace:
+```
+
+#### NetworkConfiguration Controller
+
+* User creates `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller detects whether the data entered by the
+    user is legal, and adds `finalizers`.
+
+* User deletes `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller detects whether the length of
+    `.status.networkBindingRefs` is 0, if it is not 0, it is not allowed to
+    delete the CR, otherwise the Controller deletes the `finalizers` and deletes
+    the CR.
+
+* User modify `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller deletes all the old `NetworkBinding` CRs
+    and creates a new one for each deleted `NetworkBinding` CR according to the
+    contents in `.status.networkBindingRefs`.
+
+#### NetworkBinding CRD
+
+`NetworkBinding` CR indicates a networkConfiguration is applied to some ports.
+When this CR is created, the operation of configuring network will be triggered,
+and when this CR is deleted, the operation of clearing network configuration will
+be triggered.
+This CR cannot be modified. If user wants to change the network configuration
+applied on a port, it is needed to clear the configuration first and then apply
+new configuration. So modification is done by deleting old `NetworkBinding` CR
+and creating new one.
+
+```yaml
+metaData:
+  name:
+  ownerRef:
+  finalizers:
+spec:
+  networkConfigurationRef:
+    name:
+    namespace:
+  ports:
+  - portID:
+    deviceRef:
+      name:
+      kind:
+      namespace:
+  - portID:
+    deviceRef:
+      name:
+      kind:
+      namespace:
+status:
+  state:
+```
+
+#### NetworkBinding Controller
+
+Add the environment variable `ADMIN_NAMESPACE`, the NetworkBinding Controller
+judges whether the user has the authority to configure the port according to the
+content of `Device CR annotation`.
+
+* User creates `NetworkBinding` CR
+  * Instantiate the specific `DeviceManager` according to the `deviceRef`
+    information in the `.spec.ports` field, and then call the
+    `ConfigureNetwork()` method to modify the corresponding `Device` CR.
+
+* User deletes `NetworkBinding` CR
+  * Instantiate the specific `DeviceManager` according to the `deviceRef`
+    information in the `.spec.ports` field, and then call the
+    `DeConfigureNetwork()` method to modify the corresponding `Device` CR.
+
+#### Switch
+
+```yaml
+kind: Switch
+metaData:
+  name: switch0
+  finalizers:
+  annotations:
+    adminOnlyPorts: "0/22,0/23,3/34-3/45"
+    reservedPorts: "0/1,0/2"
+spec:
+  os: fos
+  ip: 192.168.0.1
+  mac: ff:ff:ff:ff:ff:ff
+  secret:
+  # Indicates the switch at the other end of the peer link with this switch.
+  peerLinkWith:
+  - name: switch1
+    namespace: def
+status:
+  ports:
+  - portID:
+    # Current network configuration applied to this port
+    networkConfiguration:
+    # Reference refer to desired networkConfiguration CR
+    networkConfigurationRef:
+      name:
+      namespace:
```

Comment:
I don't think we need both. We can put the Ref to the current one, and update it when the port is updated, no need to keep both current and desired. At the end what matters is the desired one, and it will be overwritten if it changes. until that point it is valid.

Reply:
Our thought is that if the configuration of one port is changed, we need to first clear the previous configuration applied on that port and then apply new configuration. If only have the desired one, the cleaning will be difficult as we don't know the previous configuration. So we need to keep both current and desired here.

Reply:
I would argue that you can get the current configuration from the switch, and you anyways would want to reset everything to a default state if de-provisioning or configure everything according to the new configuration, so the old one should not matter. For example you don't care which vlan were configured when you know that you want only X and Y. you'll remove all others than X and Y.